### PR TITLE
feat(expr): support Format expression

### PR DESCRIPTION
**Pull request by TennyZhuang**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Fix https://github.com/risingwavelabs/risingwave/issues/11247

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/frontend/src/binder/expr/function.rs [hunk 1]
```diff
@@ -1032,11 +1034,80 @@ impl Binder {
...
+            .chain(formatter.nodes().iter().map(move |f_node| -> Result<_> {
+                let new_arg = match f_node {
+                    FormatterNode::Specifier(sp) => {
+                        // We've checked the count.
+                        let arg = &args[j];
+                        j += 1;
+                        match sp.ty {
+                            SpecifierType::SimpleString => arg.clone(),
+                            SpecifierType::SqlIdentifier => {
+                                FunctionCall::new(ExprType::QuoteIdent, vec![arg.clone()])?.into()
```
Comment (xiangjinwu) [inline src/frontend/src/binder/expr/function.rs:1083]:
```suggestion
                                FunctionCall::new(ExprType::QuoteIdent, vec![arg.clone().cast_output()?])?.into()
```

`quote_ident` only handles string input, while `concat_ws` and `format` handles `any`.

> Each *`formatarg`* argument is converted to text according to the usual output rules for its data type, and then formatted and inserted into the result string according to the format specifier(s).

Let's also add some tests for non-string arguments.

**Review by xiangjinwu — APPROVED** [review]
Update the PR description for doc and acknowledge the limitations:
* `formatstr` is a non-null literal
* no support for position, flags, width, `L`
* no runtime error when `%I` encounters a null value

File: src/common/src/format.rs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    /// `I` treats the argument value as an SQL identifier, double-quoting it if necessary. It is
+    /// an error for the value to be null (equivalent to `quote_ident`).
+    SqlIdentifier,
+    /// `L` quotes the argument value as an SQL literal. A null value is displayed as the string
+    /// NULL, without quotes (equivalent to `quote_nullable`).
+    SqlLiteral,
+}
+
+impl TryFrom<char> for SpecifierType {
+    type Error = ();
```
Comment (stdrc) [inline src/common/src/format.rs:32]: Why not returning `ParseFormatError` here?

**Review by stdrc — APPROVED** [review]
LGTM

File: src/common/src/format.rs (hunk 2 above)
Comment (TennyZhuang) [reply src/common/src/format.rs:32]: A part of variants are impossible here, so I don’t want to return that Error.

**Pull request merged**